#Reading the whole SILVA database header as a panda dataframe to import the complete taxonomy

import pandas as pd
from Bio import Seq, SeqIO
from Bio.Alphabet import generic_dna
import sys
import argparse

parser = argparse.ArgumentParser(description='SpeciesMLP: 16S rRNA taxonomic classifier using deep learning')
parser.add_argument('--silva_path', dest='silva_path',type=str, default='.', help='Full path of the SILVA database to parse the taxonomic ranks')
parser.add_argument('--silva_header', dest='silva_header', type=str, default='SILVA_header.csv', help='Full path of "SILVA_header.csv" a csv file contain the taxonomy ranks with ids generated by SILVA_header_2_csv.py')

# Parameters
args = parser.parse_args()

silva_path = args.silva_path
silva_header = args.silva_header


def fasta_2_df(silva_path):
	reads=[]
	for record in SeqIO.parse(silva_path, "fasta"):
	    str_ = str(record.description).split(" ", 1)[1]
	    id_ = str(record.name).split(' ')[0]
	    if str_.count(';') == 6:
	        [kingdom, phylum, class_,order,family,genus,species] = str_.split(';')
	    elif str_.count(';') > 6:    
	        list =  str_.split(';')
	        [kingdom,phylum, class_,order,family,genus,species] = list[-7:]
	    else:
	        list =  str_.split(';')
	        num = len(list)
	        [kingdom,phylum, class_,order,family,genus,species][:num] = list
	        [kingdom,phylum, class_,order,family,genus,species][num:] = '_'*(7-num)
	    reads.append([id_,kingdom ,phylum, class_,order,family,genus,species])    
	    
	SILVA_header =pd.DataFrame(reads,columns=['id','kingdom','phylum','class_','order','family','genus','species'])
	return SILVA_header

def main():
	#script = sys.argv[0]
	#silva_path = sys.argv[1]
	SILVA_header =  fasta_2_df(silva_path)
	SILVA_header.to_csv(silva_header)

if __name__ == "__main__":
    main()
